test: Update coverage.cpp to drop linux restriction

janb84 opened the pull request:
In PR [#31901](https://github.com/bitcoin/bitcoin/pull/31901), Coverage.cpp was introduced as a separate utility file, based on existing code. However, the macro defined in Coverage.cpp was limited to Clang and Linux, which caused issues for users on macOS when using the newly introduced deterministic test tooling.

This change adds fallback functions which are used when building without code coverage on non linux env.
This adds support for macOS to ResetCoverageCounters. ResetCoverageCounters is used by the unit tests in `g_rng_temp_path_init` to support the deterministic unit test tooling. It is also used in fuzz tests to completely suppress coverage from anything init-related.

See [Readme](https://github.com/bitcoin/bitcoin/blob/master/contrib/devtools/README.md) on how to test this for deterministic unit & fuzz test.  

Suggestion for test files: 

- for  unit test: `util_string_tests`
- for fuzz test: `addition_overflow `

These files should give deterministic results

Sjors commented:
From the original PR it's not obviously how to use this, is there a tl&dr for what to test?

janb84 commented:
> From the original PR it's not obvious how to use this, is there a tl&dr for what to test?

Use this config:
```shell
cmake -S . -B build -DCMAKE_C_COMPILER='clang' -DCMAKE_CXX_COMPILER='clang++' -DCMAKE_CXX_FLAGS='-fprofile-instr-generate -fcoverage-mapping'
```
build as normal :
```shell
 cmake --build build 
 ```
 Run tooling for util_string_tests:
 ```shell
 RUST_BACKTRACE=1 cargo run --manifest-path ./contrib/devtools/deterministic-unittest-coverage/Cargo.toml -- $PWD/build util_string_tests
 ````
 
 Should return :
 ```shell
 *** No errors detected
The coverage was deterministic across two runs.
```

brunoerg commented on src/test/util/coverage.cpp:
```diff
@@ -4,7 +4,7 @@
 #include <test/util/coverage.h>
 
-#if defined(__clang__) && defined(__linux__)
+#if defined(__clang__) && (defined(__linux__) || defined(__APPLE__))
```
I think linux is more permissive than macOS that's why CI is failing. The jobs do not set `-fprofile-instr-generate -fcoverage-mapping`. I did not test, but using lld could work?

janb84 replied:
Thank you for the suggestion, have found a way to fix the code so that the linker does not break and the functionality is the same. Thank you for the review

Prabhat1308 commented:
tACK [611151e](https://github.com/bitcoin/bitcoin/pull/32059/commits/611151eae75ed756437442a5d0356f525c73dd00)

fanquake commented on src/test/util/coverage.cpp:
```diff
@@ -4,19 +4,18 @@
 #include <test/util/coverage.h>
 
-#if defined(__clang__) && defined(__linux__)
+#if defined(__clang__) && (defined(__linux__) || defined(__APPLE__))
```
Why do we need to specify platforms here, rather than just `__clang__` (especially if you're now going to add fallbacks)?

janb84 replied:
Good point, I have removed the check in the macro and rebased my commits. Thank you for the review.

janb84 commented on src/test/util/coverage.cpp:
```diff
@@ -8,15 +8,14 @@
-        __llvm_profile_reset_counters();
-    }
+// Fallback implementations
+extern "C" void __llvm_profile_reset_counters(void) {}
```
Rebased the cod and changed the PR description based on feedback , thank you for the review!

janb84 commented:
> > See [Readme](https://github.com/bitcoin/bitcoin/blob/master/contrib/devtools/README.md) on how to test this for deterministic unit test
> 
> I guess this can be used to test both (unit and fuzz tests)? If yes, it could make sense to recommend a unit test and fuzz test, each. For example `util_string_tests` and `addition_overflow` (haven't tested either)? Before, the coverage should be non-deterministic, and after this change, it should be deterministic.

Have incorporated your suggestion and checked that the files are indeed deterministic (they are). Again thanks for the review

hodlinator commented on src/test/util/coverage.cpp:
```diff
@@ -4,19 +4,18 @@
 
-void ResetCoverageCounters()
-{
-    if (__llvm_profile_reset_counters) {
-        __llvm_profile_reset_counters();
-    }
+// Fallback implementations
+extern "C" void __llvm_profile_reset_counters(void) {}
+extern "C" void __gcov_reset(void) {}
```
Not sure I've ever come across this corner of C/C++ before. So you define `__llvm_profile_reset_counters(void) __attribute__((weak))` for it to hopefully link to something external, but then define local implements for them in case we don't link to code coverage libs (build with code coverage enabled)? So for when we include this test-related file but only include unit tests, and build without coverage support?

From the PR description:

> This change extends the functionality of the function with fallback code so that the linker does not fail on non-linux env.

Wonder if it would be more correct to say:

> This change adds fallback functions which are used when building without code coverage.

?

janb84 replied:
### Main problem: 

> "macOS’s linker (ld64) doesn’t silently resolve unreferenced weak symbols to NULL when they’re called—it demands a definition".  

Combined with calling `__gcov_reset` and `__llvm_profile_reset_counters` directly, forcing the linker to find a definition, will result in that the linker will break on trying to link the following code: 
```C
extern "C" void __llvm_profile_reset_counters(void) __attribute__((weak));
extern "C" void __gcov_reset(void) __attribute__((weak));
```
Solution is to add a fallback option :
```C
extern "C" __attribute__((weak)) void __llvm_profile_reset_counters(void) {}
extern "C" __attribute__((weak)) void __gcov_reset(void) {}
```
These provide weak definitions—empty implementations that the linker can use if no stronger definitions exist.

With this in place there, are 2 scenario's

### 1 When Profiling Is Disabled

- Weak definitions are the only ones available. The linker doesn’t need to choose between multiple weak definitions or rely on order—it picks the {} one because they’re the sole candidates.
- The weak attribute ensures no conflict if strong definitions appear later (e.g., with profiling enabled), but without profiling, the weak {} versions are sufficient to satisfy the linker.
- This avoids the “undefined symbol” error as seen before on macOS because a definition exists, even if weak.

### 2 When Profiling Is Enabled 

- The runtime provides strong definitions.
- The weak definitions are overridden (strong beats weak).
- ResetCoverageCounters() calls the runtime’s implementations.

hodlinator replied:
Thanks for the thorough explanation, it was helpful!

nit:
This seems to build fine on Linux with/without coverage:
```C++
extern "C" void __llvm_profile_reset_counters(void);
extern "C" void __gcov_reset(void);
```
My guess is that what is important now is that the fallback is `weak`, not the prototype, could that be? Mac CI seems to accept that, but I'm not sure if our CI tests both non-coverage and coverage builds on Mac.

---
Iff the above is a problem on Mac/other platforms though, I have a different variant to suggest:
Others might disagree, but I would be happy if you added a second commit which just refactors these existing prototypes to be consistent with their fallback implementations.
```C++
extern "C" __attribute__((weak)) void __llvm_profile_reset_counters(void);
extern "C" __attribute__((weak)) void __gcov_reset(void);
```
Makes them line up visually with the new fallbacks, and this is test-related code, so we're a bit more lenient on touching unrelated lines.

It's also consistent with [Clang docs](https://clang.llvm.org/docs/UsersManual.html#fine-tuning-profile-collection):
```C++
__attribute__((weak)) int __llvm_profile_dump(void);
```

(I attempted the alternative of instead changing the fallback lines to have `__attribute__((weak))` after the function args but Clang warns about GCC incompatibility).

hodlinator replied:
Thanks for lining up the attributes!

janb84 commented:
In my write-up of the explanation to hodlinator I discovered one edge case that needed fixing, sorry for the new commit.

Reason: 
Weak fallbacks (__attribute__((weak))) were added instead of strong ones to ensure the profiling runtime’s implementations are used when profiling is enabled, **_avoiding linker conflicts or silent overrides_**. This guarantees correct counter resets while keeping safe, empty fallbacks when profiling is off, improving portability and reliability across Clang and GCC.

hodlinator reviewed (approved):
ACK ecdbe72916c1cabbf810e892d25fe6eed30b1306

Tested on Linux by running the [recommended](https://github.com/bitcoin/bitcoin/pull/32059#issuecomment-2722068136) Rust deterministic unit test coverage utility both without and with coverage support, with expected results.

janb84 committed:
test: Enable ResetCoverageCounters beyond Linux

Non-Linux linkers require a fallback implementation for when coverage is not enabled.
The fallbacks are marked weak to have lower precedence than built-in implementations when available, removing ambiguity from the linker.

janb84 commented on src/test/util/coverage.cpp:
Ah thanks! Have changed the commit message accordingly.

hodlinator reviewed (approved):
re-ACK 54e6eacc1fccd602897d9e3025c62f83194ffd5b

Changes since [first ACK](https://github.com/bitcoin/bitcoin/pull/32059#pullrequestreview-2686973172):
- Moved `weak` attribute on the original declarations so they line up with the fallback definitions and mirror [Clang docs](https://github.com/bitcoin/bitcoin/pull/32059#discussion_r1997328341). Thanks!
- Improved commit message.

Retested using Rust utility on coverage-enabled version of the code with successful determinism result.
